src/omv/ports/mimxrt/omv_spi.c

```diff
@@ -78,65 +107,118 @@ static void spi_master_callback(LPSPI_Type *base, void *handle, status_t status,
-    if ((spi->xfer_flags & flags) == flags) {
-        // Restart transfer for circular transfers.
-        spi->xfer_flags &= ~(OMV_SPI_XFER_COMPLETE);
-        LPSPI_MasterTransferEDMA(spi->inst, &spi->descr_master_edma, &spi->xfer_descr);
+    // Clear after the callback so the callback gets the xfer complete flag set.
+    // This needs to be cleared to prevent circular DMA from re-triggering on a failure.
+    spi->xfer_flags &= ~(OMV_SPI_XFER_COMPLETE);
+}
+
+// The LPSPI_MasterTransferNonBlocking function call does not clear the LPSPI_TCR_FRAMESZ_MASK bit
+// even if kLPSPI_MasterByteSwap is not set in the xfer. So, we have to clear this bit at the same
+// time as changing the framesize. Splitting this will result in inconsitent settings. Finally,
+// the TCR register doesn't always update itself correctly so you have to try in a loop...
+void omv_spi_set_frame_size(omv_spi_t *spi, uint16_t size) {
+    for (;;) {
+        uint32_t tcr = spi->inst->TCR;
+        uint16_t s = (uint16_t) ((tcr & LPSPI_TCR_FRAMESZ_MASK) >> LPSPI_TCR_FRAMESZ_SHIFT) + 1U;
+        if (s != size) {
+            spi->inst->TCR = (tcr & ~(LPSPI_TCR_BYSW_MASK | LPSPI_TCR_FRAMESZ_MASK)) | LPSPI_TCR_FRAMESZ(size - 1U);
+        } else {
+            break;
+        }
     }
 }
 
 int omv_spi_transfer_start(omv_spi_t *spi, omv_spi_transfer_t *xfer) {
     spi->callback = xfer->callback;
     spi->userdata = xfer->userdata;
+    spi->xfer_error = 0;
     spi->xfer_flags = xfer->flags;
-    // Duplicate & stash transfer in spi struct, to avoid recreating
-    // it for circular transfers. NOTE: If circular transfers ever
-    // works this can be removed.
+
+    uint32_t datasize = 8;
+    if (xfer->flags & OMV_SPI_XFER_16_BIT) {
+        datasize = 16;
+    }
+
+    omv_spi_set_frame_size(spi, datasize);
+
     spi->xfer_descr.txData = xfer->txbuf;
     spi->xfer_descr.rxData = xfer->rxbuf;
     spi->xfer_descr.dataSize = xfer->size;
-    spi->xfer_descr.configFlags = kLPSPI_MasterPcs0 | kLPSPI_MasterPcsContinuous | kLPSPI_MasterByteSwap;
+    spi->xfer_descr.configFlags = kLPSPI_MasterPcs0 | kLPSPI_MasterPcsContinuous;
     spi->xfer_flags &= ~(OMV_SPI_XFER_FAILED | OMV_SPI_XFER_COMPLETE);
 
-    if (spi->xfer_flags & OMV_SPI_XFER_DMA) {
-        // DMA transfer (circular or one-shot)
-        if (LPSPI_MasterTransferEDMA(spi->inst, &spi->descr_master_edma, &spi->xfer_descr) != kStatus_Success) {
+    if (xfer->flags & OMV_SPI_XFER_16_BIT) {
+        spi->xfer_descr.dataSize *= 2;
+    }
+
+    if (spi->dma_flags & OMV_SPI_DMA_DOUBLE) {
+        spi->xfer_descr.dataSize /= 2;
+    }
+
+    // Check if we can optimize the transfer by doing things in 32-bit chunks.
+    if ((!(spi->xfer_descr.dataSize % 4)) && (!(xfer->flags & OMV_SPI_XFER_16_BIT))) {
+        omv_spi_set_frame_size(spi, 32);
+        spi->xfer_descr.configFlags |= kLPSPI_MasterByteSwap;
+    }
+
+    if (spi->xfer_flags & OMV_SPI_XFER_BLOCKING) {
```

Comment:
Please revert this to the original order it used, DMA was checked first, it will make it easier to read the diff.

Reply:
Okay.